### PR TITLE
airbyte-ci: accept extra options on `connectors test` and pass these to test steps

### PR DESCRIPTION
**Pull request by alafanechere**
## What
Relates to https://github.com/airbytehq/airbyte/issues/24061
In https://github.com/airbytehq/airbyte/pull/34049 I updated the `Step` class to receive extra parameters and use them.
In this PR  I'm updating the `connectors test` command to handle step specific extra parameters and pass them to the Step.

## How
1.  Add unprocessed arguments to the `test` comment to pass arbitrary options that we'll consider being extra parameters for steps
2. Parse and validate these extra parameters 
3. Pass them to the Steps.

E.G:
Passing a different option to python unit tests:
```
airbyte-ci --disable-dagger-run connectors --name=source-faker test --unit.--cov-fail-under=100
````

## Caveats / Limitation

* As some step id are mapped to different Step class we can't guarantee that an extra parameter will be correctly handled during step execution. Example: I run the  `test` command on a python and a java connectors with `--unit.cov-fail-under=100` option. The `cov-fail-under` option only exists in the python connector land (it's a pytest option) , so the gradle test for the java connector will likely fail as the command  with this specific option is invalid.

## TODO
Add unit tests in a different branch

### CODE REVIEW

**Comment by alafanechere** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/34091?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

<b>In this stack:</b> <code>🔧 Enhance connector testing with extra parameters</code>
* `master`
  * **#34050** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/34050?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **#34049** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/34049?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **#34091** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/34091?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @alafanechere and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py [hunk 1]
```diff
@@ -23,14 +34,19 @@
 }
 
 
+EXTRA_PARAMS_PATTERN = re.compile(r"^--([a-zA-Z_][a-zA-Z0-9_]*)\.([a-zA-Z_-][a-zA-Z0-9_-]*)=([^=]+)$")
```
Comment (bnchrch) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py:37]: 😍 Love the regex
📚Lets make sure we give an example of what should / shouldnt match

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py [hunk 2]
```diff
@@ -39,27 +55,105 @@ def get_test_steps(context: ConnectorContext) -> STEP_TREE:
         return []
 
 
-async def run_connector_test_pipeline(context: ConnectorContext, semaphore: anyio.Semaphore) -> ConnectorReport:
+def build_extra_params_mapping(
+    extra_params: Tuple[str], unique_steps_to_run: Dict[CONNECTOR_TEST_STEP_ID, Step]
```
Comment (bnchrch) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py:59]: ❓Is extra params a tuple or a list?

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py [hunk 3]
```diff
@@ -39,27 +55,105 @@ def get_test_steps(context: ConnectorContext) -> STEP_TREE:
...
+                f"Invalid step name {step_name}, it must be one of {[step_id.value for step_id in unique_steps_to_run.keys()]}"
+            )
+        step_id = CONNECTOR_TEST_STEP_ID(step_name)
+        if step_id not in extra_params_mapping:
+            extra_params_mapping[step_id] = {}
+        if param_name not in extra_params_mapping[step_id]:
+            extra_params_mapping[step_id][param_name] = []
+        if param_value is not None:
+            extra_params_mapping[step_id][param_name].append(param_value)
+    return extra_params_mapping
```
Comment (bnchrch) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py:103]: 💡Could use `setdefault`
```suggestion
        step_id = CONNECTOR_TEST_STEP_ID(step_name)
        extra_params_mapping.setdefault(step_id, {}).setdefault(param_name, []).append(param_value)

    return extra_params_mapping
```

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py [hunk 4]
```diff
@@ -39,27 +55,105 @@ def get_test_steps(context: ConnectorContext) -> STEP_TREE:
...
+
+
+def log_steps_params(unique_steps_to_run: Dict[CONNECTOR_TEST_STEP_ID, Step]) -> None:
```
Comment (bnchrch) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py:106]: 😍

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py [hunk 5]
```diff
@@ -39,27 +55,105 @@ def get_test_steps(context: ConnectorContext) -> STEP_TREE:
...
             [
                 StepToRun(id=CONNECTOR_TEST_STEP_ID.METADATA_VALIDATION, step=MetadataValidation(context)),
                 StepToRun(id=CONNECTOR_TEST_STEP_ID.VERSION_FOLLOW_CHECK, step=VersionFollowsSemverCheck(context)),
                 StepToRun(id=CONNECTOR_TEST_STEP_ID.VERSION_INC_CHECK, step=VersionIncrementCheck(context)),
                 StepToRun(id=CONNECTOR_TEST_STEP_ID.QA_CHECKS, step=QaChecks(context)),
             ]
         ]
+        all_steps_to_run += static_analysis_steps_to_run
+
+    set_extra_params_on_steps_to_run(extra_params, all_steps_to_run)
```
Comment (bnchrch) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py:151]: ❗This method is best placed inside `run_steps`. As this is the area responsible for evaluating the step tree at run time, and extra_args are a runtime concern.

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py [hunk 6]
```diff
@@ -39,27 +55,105 @@ def get_test_steps(context: ConnectorContext) -> STEP_TREE:
...
+        if is_flag:
+            step_name, param_name = matches.groups()
+            param_value = None
+        else:
+            step_name, param_name, param_value = matches.groups()
+        if step_name not in valid_step_names or CONNECTOR_TEST_STEP_ID(step_name) not in unique_steps_to_run:
+            raise ValueError(
+                f"Invalid step name {step_name}, it must be one of {[step_id.value for step_id in unique_steps_to_run.keys()]}"
+            )
+        step_id = CONNECTOR_TEST_STEP_ID(step_name)
```
Comment (bnchrch) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py:96]: ❗Given that `extra_params` is a useful concept for all pipelines we should avoid linking its parsing logic to connector tests directly.

We could either
1. Not worry if they pass a not found id (as the arg will not apply anywhere)
2. Or raise this during `run_steps` as "warning: no corresponding step found for argument in run tree"

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/commands.py [hunk 7]
```diff
@@ -47,13 +53,15 @@
     type=click.Choice([step_id.value for step_id in CONNECTOR_TEST_STEP_ID]),
     help="Skip a step by name. Can be used multiple times to skip multiple steps.",
 )
+@click.argument("extra_params", nargs=-1, type=click.UNPROCESSED)
```
Comment (bnchrch) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/commands.py:56]: ❓ Should we combine this decorator and `build_extra_params_mapping` into a custom decorator

```
@allow_extra_params
```

That will parse the raw `extra_params` given by click into a usable dict?

**Review by bnchrch — COMMENTED** [review]
This is looking really good! left a few broad questions.

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py (hunk 2 above)
Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py:59]: It's a tuple of strings when received from the CLI, this function transforms it  to a dict

**Comment by alafanechere** [comment]
@bnchrch thanks for this thoughtful comments!
I made the following changes which I think matches what you suggested:
* The `extra_params` parsing is handled at the entrypoint level. It's only failing if the regex does not match and the step id is not one present in the enum
* I pass the parsed extra params to the `RunStepOptions`
* Setting extra params on Step happens in `run_step`, consuming step specific options from the `RunStepOptions`

I'd prefer not using an additional decorator `@allow_extra_params` as we lose the explicitness that `extra_params` are additional command arguments.

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/argument_parsing.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        matches = pattern.match(param)
+        if not matches:
+            raise ValueError(f"Invalid parameter {param}. {EXTRA_PARAM_PATTERN_ERROR_MESSAGE}")
+        if is_flag:
+            step_name, param_name = matches.groups()
+            param_value = None
+        else:
+            step_name, param_name, param_value = matches.groups()
+        try:
+            step_id = CONNECTOR_TEST_STEP_ID(step_name)
```
Comment (bnchrch) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/argument_parsing.py:57]: I would advocate that we keep this function generic and unrelated to connector tests as extra args will likely be used in many other pipelines

Perhaps we instead
1. Move this function near where we define `run_steps`
2. Pass in a list of valid step ids and use them in place of `CONNECTOR_TEST_STEP_ID`

**Review by bnchrch — CHANGES_REQUESTED** [review]

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/argument_parsing.py (hunk 8 above)
Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/argument_parsing.py]: @bnchrch fair enough! I made the function more generic. Should be better now.

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/argument_parsing.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+# Pattern  for extra param options: --<step_id>.<option_name>=<option_value>
+EXTRA_PARAM_PATTERN_FOR_OPTION = re.compile(r"^--([a-zA-Z_][a-zA-Z0-9_]*)\.([a-zA-Z_-][a-zA-Z0-9_-]*)=([^=]+)$")
+# Pattern  for extra param flag: --<step_id>.<option_name>
+EXTRA_PARAM_PATTERN_FOR_FLAG = re.compile(r"^--([a-zA-Z_][a-zA-Z0-9_]*)\.([a-zA-Z_-][a-zA-Z0-9_-]*)$")
+EXTRA_PARAM_PATTERN_ERROR_MESSAGE = "The extra flags must be structured as --<step_id>.<flag_name> for flags or --<step_id>.<option_name>=<option_value> for options. You can use - or -- for option/flag names."
+
+
+def build_extra_params_mapping(
+    SupportedStepId: Type[Enum]
```
Comment (bnchrch) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/argument_parsing.py:23]: nit: we should pluralize this. (also I know this is my original mistake, sorry!)
```suggestion
    SupportedStepIds: Type[Enum]
```

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/commands.py [hunk 10]
```diff
@@ -3,22 +3,30 @@
...
 import sys
-from typing import List
+from typing import Dict, List
 
 import asyncclick as click
 from pipelines import main_logger
 from pipelines.airbyte_ci.connectors.consts import CONNECTOR_TEST_STEP_ID
 from pipelines.airbyte_ci.connectors.context import ConnectorContext
 from pipelines.airbyte_ci.connectors.pipeline import run_connectors_pipelines
+from pipelines.airbyte_ci.connectors.test import argument_parsing
```
Comment (bnchrch) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/commands.py:13]: ❗Last thing in the journey to being generic. Lets move this to a generic location

```suggestion
from pipelines.airbyte_ci.helpers import argument_parsing
```

or better yet create a new area that denotes "These are things we use to execute the step tree"


```suggestion
from pipelines.helpers.execution.argument_parsing  import build_extra_params_mapping
from pipelines.helpers.execution.run_steps import RunStepOptions
```

**Review by bnchrch — APPROVED** [review]
All is left (imo) is moving a file to a new path.

Going to preapprove so I can be sure I am out of your way monday!

[looksgood](https://app.graphite.dev/macros/airbytehq)

**Pull request merged**